rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,95 @@
+- It should also cover any alterations that might need to be made to the expected data format
+
+## Motivation
+
+The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing a proprietary format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.
+
+## Internal Proposal
+
+From a high level my proposal is pretty simple. I'll dive into the rationale and context lower in the doc but lets just tee that up, shall we?
+
+I propose that if we're not searching explicitly for better perf around these messages and we're not explicitly searching for alternate tooling because we've decided that we wholly can't deal with the tooling overhead of protobufs we instead continue to use them with prost. I would also posit that if the concerns around exposing our private data format are bothersome enough, we should consider following an open spec for the data as a primary example: [opentelemetry](https://opentelemetry.io/).
+
+Should performance be a heavier concern for this standardization than I have assumed for the writing of this doc, my propsal would be to consider a highly performant alternative to protobufs like Flatbuffers and still consider following an open spec for the data. The `Rationale` section below will cover my thoughts on this more thoroughly.
+
+## Doc-level Proposal
+
+I'm fairly certain that this should not require a docs entry as it's largely internal to the tools. There is a possibility that we might want to document the data format we use if either to dissuade users from following it/using it (in the case that we don't adopt a standardized data format) or to point out that we've standardized on an open data spec that can be followed and used in other areas.
+
+## Rationale
+
+Unfortunately in most of my research I've pretty well come to the conclusion that encoding formats as a domain are rife with technical and political failures. That is to say, in some regard or another, every format from JSON to Flatbuffers is going to force a tradeoff between correctness, performance, features or tooling and maintenance overhead. Inlined below here is a reasonable starting point on performance. It's taken from the [NoProto project](https://https://github.com/only-cliches/NoProto) since they're regularly updating their benchmarks. They've also done an excellent job of enumerating some of the aspects to consider when making a decision of this nature and their read on the space very much mirrors my own thinking in many ways.
```

Comment:
```suggestion
Unfortunately in most of my research I've pretty well come to the conclusion that encoding formats as a domain are rife with technical and political failures. That is to say, in some regard or another, every format from JSON to Flatbuffers is going to force a tradeoff between correctness, performance, features or tooling and maintenance overhead. Inlined below here is a reasonable starting point on performance. It's taken from the [NoProto project](https://github.com/only-cliches/NoProto) since they're regularly updating their benchmarks. They've also done an excellent job of enumerating some of the aspects to consider when making a decision of this nature and their read on the space very much mirrors my own thinking in many ways.
```